depthmapX/gltrianglesuniform.cpp

```diff
@@ -0,0 +1,168 @@
+    "}\n";
+
+static const char *fragmentShaderSource =
+    "uniform vec4 colourVector;\n"
+    "void main() {\n"
+    "   gl_FragColor = colourVector;\n"
+    "}\n";
+
+/**
+ * @brief GLPolygon::GLPolygon
+ * This class is an OpenGL representation of a set of triangles of uniform colour
+ */
+
+GLTrianglesUniform::GLTrianglesUniform()
+    : m_count(0),
+      m_program(0)
+{
+
+}
+
+void GLTrianglesUniform::loadTriangleData(const std::vector<Point2f>& points, const PafColor &polyColour)
+{
+    built = false;
+
+    m_count = 0;
+    m_data.resize(points.size() * DATA_DIMENSIONS);
+
+    std::vector<Point2f>::const_iterator iter = points.begin(), end =
+    points.end();
+    for ( ; iter != end; ++iter )
+    {
+        const Point2f & point = *iter;
+        add(QVector3D(point.x, point.y, 0.0f));
+    }
+    m_colour.setX(polyColour.redf());
+    m_colour.setY(polyColour.greenf());
+    m_colour.setZ(polyColour.bluef());
+}
+
+void GLTrianglesUniform::setupVertexAttribs()
+{
+    m_vbo.bind();
+    QOpenGLFunctions *f = QOpenGLContext::currentContext()->functions();
+    f->glEnableVertexAttribArray(0);
+    f->glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, DATA_DIMENSIONS * sizeof(GLfloat), 0);
+    m_vbo.release();
+}
+
+void GLTrianglesUniform::initializeGL(bool m_core)
+{
+    if(m_data.size() == 0) return;
+    m_program = new QOpenGLShaderProgram;
+    m_program->addShaderFromSourceCode(QOpenGLShader::Vertex, m_core ? vertexShaderSourceCore : vertexShaderSource);
+    m_program->addShaderFromSourceCode(QOpenGLShader::Fragment, m_core ? fragmentShaderSourceCore : fragmentShaderSource);
+    m_program->bindAttributeLocation("vertex", 0);
+    m_program->link();
+
+    m_program->bind();
+    m_projMatrixLoc = m_program->uniformLocation("projMatrix");
+    m_mvMatrixLoc = m_program->uniformLocation("mvMatrix");
+    m_colourVectorLoc = m_program->uniformLocation("colourVector");
+
+    m_vao.create();
+    QOpenGLVertexArrayObject::Binder vaoBinder(&m_vao);
+
+    m_vbo.create();
+    m_vbo.bind();
+    m_vbo.allocate(constData(), m_count * sizeof(GLfloat));
+
+    setupVertexAttribs();
+    m_program->setUniformValue(m_colourVectorLoc, m_colour);
+    m_program->release();
+    built = true;
+}
+
+void GLTrianglesUniform::updateGL(bool m_core) {
+    if(m_program == 0) {
+        // has not been initialised yet, do that instead
+        initializeGL(m_core);
+    } else {
+        m_vbo.bind();
+        m_vbo.allocate(constData(), m_count * sizeof(GLfloat));
+        m_vbo.release();
+        built = true;
+    }
+}
+
+void GLTrianglesUniform::updateColour(const PafColor &polyColour)
+{
+    m_colour.setX(polyColour.redf());
+    m_colour.setY(polyColour.greenf());
+    m_colour.setZ(polyColour.bluef());
+    m_program->bind();
+    m_program->setUniformValue(m_colourVectorLoc, m_colour);
+    m_program->release();
+}
+
+void GLTrianglesUniform::cleanup()
+{
+    if(!built) return;
```

Comment:
Yeah, it's called in the [destructor of GLView](https://github.com/SpaceGroupUCL/depthmapX/pull/67/files/b4aac446ca9a02bd387ee7cc9cbc5b937dea2724#diff-2975c182b14becadce581dc3f0e38914R69) because it requires the openGL context (from makeCurrent()). There doesnt seem to be safer way...